### ISSUE TITLE
Improve performance of CosmosSerializer

### ISSUE THREAD

**Issue by nahk-ivanov** [Azure/azure-cosmos-dotnet-v3#1183]
**Is your feature request related to a problem? Please describe.**
Current APIs of `CosmosSerializer` are rather strange. There are two main issues:
1) When working with streams it is assumed to by I/O, so methods should be asynchronous.
2) `ToStream` expects you to *return* a stream, instead of providing a stream to write to. I assume the contents of the returned stream are later copied to the actual request stream, but this is completely unnecessary. With the current implementation it would be more suitable to make it return a `byte[]` or something, but ideally you want to be provided a stream to serialize to.

I also noticed that default implementation (`CosmosJsonDotNetSerializer`) is closing the provided stream in `FromStream`, which makes me think that it receives a copy of the contents from the actual response stream. For performance reasons it's better to provide the actual response stream and deserialize from it straight. With synchronous APIs you probably want to read entire response stream into memory and pass a `Span`/`Buffer` to the `CosmosSerializer.FromStream(..)`.

**Describe the solution you'd like**
1) Both `To/FromStream` methods should be asynchronous.
2) `ToStream` should provide a `Stream` instance to write data to.

**Describe alternatives you've considered**
None. I just had to implement it synchronously and (unnecessarily) allocate `MemoryStream`, just to wrap `byte[]` into it.

**Additional context**
Current implementation does not really align well with built-in `System.Text.Json` serialization and mostly geared towards old and slow JSON.NET. Given that new ASP.NET Core infrastructure relies on `System.Text.Json` by default, it will be beneficial if `CosmosSerializer` APIs would be updated for performance.

**Comment by j82w** [comment]
@kirankumarkolli can you explain this behavior?

**Comment by j82w** [comment]
@nahk-ivanov the main reason we went with a synchronous API is newtonsoft which is used as the default serializer doesn't have an async serialization methods and System.Text.Json was still in preview when v3 went GA.

**Comment by nahk-ivanov** [comment]
It seems like [v4](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/d69794ad591e8eaafa4bc811387ccee35b4f6d78/Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializer.cs) has the same problems at the moment, why close the issue?

Do you disagree that stream I/O should be async? Do you disagree that `ToStream` should write to stream, rather than return one? I'd suggest we keep the issue open until we actually address this or agree that current implementation is objectively better (which, honestly, I don't even see as a debate).

**Comment by j82w** [comment]
Sorry, I thought v4 was converted to be Async. 

@ealsur any reason v4 isn't using async for the serializer?

**Comment by j82w** [comment]
For v3 Netwonsoft doesn't offer an async version. Making the method async would just cause a performance hit by forcing all the calls to be wrapped in an unnecessary task.

Passing a stream in to write to vs return one is interesting. The current design was based on the fact it gives users the most customizability. A custom serializer can implement a custom pool for streams and decide on how to optimize it. If the SDK passes in a stream then it forces users to always use it's allocation system.

**Comment by nahk-ivanov** [comment]
> For v3 Netwonsoft doesn't offer an async version. Making the method async would just cause a performance hit by forcing all the calls to be wrapped in an unnecessary task.

You should implement it synchronously and return `Task.FromResult` at the end - that way there is little overhead. You are designing an SDK here, not an application. You should think about extensible, rather than basing it on the default implementation.

> The current design was based on the fact it gives users the most customizability.

It's funny to see performance and customization mentioned in one reply, like you took two completely different paths in one SDK which made it inconsistent with itself.
Are there any scenarios where serialized JSON is not being sent to the server by the SDK? I don't think there is any optimization possible that would make it run faster, than writing to the request stream directly.

**Comment by ealsur** [comment]
V4 does not use Async. Newtonsoft originally had Async versions but later deprecated them in favor of Sync only.
For V4 we investigated how System.Text.Json does Stream serialization and the part that deals with the Stream itself, it's actually Sync, see  https://github.com/dotnet/runtime/blob/master/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs#L577

So when leveraging System.Text.Json, adding Async versions had really not improvement as there was no Async operations involved.

**Comment by ealsur** [comment]
Also, this is the System.Text.Json serializer version https://github.com/Azure/azure-cosmos-dotnet-v3/blob/1e6a86c50d70520deadb4a5dd07df878b3fdc138/Microsoft.Azure.Cosmos/azuredata/Serializer/CosmosTextJsonSerializer.cs

**Comment by nahk-ivanov** [comment]
> So when leveraging System.Text.Json, adding Async versions had really not improvement as there was no Async operations involved.

I don't think this is a valid justification. Even if it doesn't do it asynchronously today does not mean it will not do so tomorrow. You are running into the same problem as the current implementation - trying to make assumptions on the current underlying components you use, rather than being future-proof. Users can have their own implementation of the `CosmosSerializer` that will work with streams asynchronously, but it is not possible in V3.

**Comment by ealsur** [comment]
> 
> 
> > So when leveraging System.Text.Json, adding Async versions had really not improvement as there was no Async operations involved.
> 
> I don't think this is a valid justification. Even if it doesn't do it asynchronously today does not mean it will not do so tomorrow. You are running into the same problem as the current implementation - trying to make assumptions on the current underlying components you use, rather than being future-proof. Users can have their own implementation of the `CosmosSerializer` that will work with streams asynchronously, but it is not possible in V3.

That is a fair point, the question though is, how does the SDK know which API to use when internally deserializing the Stream to the class? Should it use the Async version or the Sync one? Should we only have an Async one? But then, we would be forcing everyone that uses sync to wrap on Task.FromResult.

**Comment by ealsur** [comment]
There are two Stream interactions. One is Service Response -> User type, this is purely CPU bound, there is no I/O, since it's a MemoryStream. The other one is User Type -> Service request, in this case, it could be I/O bound if the User Type involves I/O access.

In the first case, since there is no I/O, which would be the perf benefit of FromStream Async?
In the later case, since we expose Stream APIs (like CreateItemStreamAsync) which already take a Stream, you can do your I/O bound async work before calling the method. For non I/O bound scenarios, which would be the perf benefit of the Async ToStream?

**Comment by kirankumarkolli** [comment]
Current ComsosSerializer contract do result in at-least one extra copy. Buffering/full materialization did help in some contexts around retries, reliability etc... but mainline cases can avoid it.  

Fully wiring it E2E avoiding copy is a quite expensive refactoring.
Exploring MemoryStream or Byte[] as contracts for V4 definitely makes sense. 
@ealsur please track it for V4.

**Comment by j82w** [comment]
@kirankumarkolli   [Memory<T>](https://docs.microsoft.com/en-us/dotnet/standard/memory-and-spans/memory-t-usage-guidelines) is another option that might have better performance over Byte[].

**Comment by nahk-ivanov** [comment]
> how does the SDK know which API to use when internally deserializing the Stream to the class? Should it use the Async version or the Sync one? Should we only have an Async one? But then, we would be forcing everyone that uses sync to wrap on Task.FromResult.

It should be async only and there should be a base class that simplifies synchronous implementations for those who want to block a thread for network transmission (do these people still exist in 21-st century?). Standard stuff, don't reinvent the wheel.
```csharp
interface ICosmosSerializer
{
    Task<T> FromStreamAsync<T>(Stream input);
}

abstract class SynchronousCosmosSerializer
{
    public Task<T> FromStreamAsync<T>(Stream input)
    {
        return Task.FromResult(FromStream(input));
    }

    protected abstract T FromStream<T>(Stream input);
}
```

> There are two Stream interactions. One is Service Response -> User type, this is purely CPU bound, there is no I/O, since it's a MemoryStream. The other one is User Type -> Service request, in this case, it could be I/O bound if the User Type involves I/O access.

Both cases have I/O and are bound on network more than CPU most likely, thus should be async. The fact that "it's a MemoryStream" is part of the problem - we don't need to read and buffer the full response body to deserialize. I could not make sense of the rest of the comment, it seems like you don't consider network (client-server request/response) an I/O operations and only think that writing to locally attached devices incurs the cost. In fact, network I/O is way slower than writing to local SSD, so I'd be more worried about parsing the response from a socket or writing the request to the socket be async and available to custom serializer, rather than whatever I do with the JSON afterwards.
You can, of course, ask `CosmosSerializer` to deserialize from a buffer and serialize into a buffer, while doing actual read and write from network async, but this results in unnecessary memory consumption, that I think can (and should) be avoided.
The overall paradigm is shifting towards `async` throughout the whole program: there is `IAsyncDisposable`, async enumerables and `ValueTask` actually allows it with very little overhead. There is no reason for something that can be `async` not to be `async` and synchronous implementation is an outlier.

@kirankumarkolli I don't know what's expensive and what's not, guys can decide based on your velocity. What I'm saying we have a questionable design in v3 that could use some improvement and since you are working on v4, where backwards compatibility might be broken in a favor of better perf - we may consider actually doing it.

@j82w, if we want to stick with synchronous implementation, yeah. In my mind, we should be able to avoid `Stream.ReadToEnd` completely, if we work over the in/out streams directly in `async` fashion.
